src/mame/drivers/gradius3.cpp

```diff
@@ -165,23 +122,25 @@ void gradius3_state::gradius3_map(address_map &map)
 	map(0x0e0000, 0x0e0001).w("watchdog", FUNC(watchdog_timer_device::reset16_w));
 	map(0x0e8000, 0x0e8000).w("soundlatch", FUNC(generic_latch_8_device::write));
 	map(0x0f0000, 0x0f0001).w(this, FUNC(gradius3_state::sound_irq_w));
-	map(0x100000, 0x103fff).ram().share("share1");
-	map(0x14c000, 0x153fff).rw(this, FUNC(gradius3_state::k052109_halfword_r), FUNC(gradius3_state::k052109_halfword_w));
-	map(0x180000, 0x19ffff).ram().w(this, FUNC(gradius3_state::gradius3_gfxram_w)).share("k052109");
+	map(0x100000, 0x103fff).ram().share("shared_ram");
+	/* is this a bug in the game or something else? */
+	map(0x14c000, 0x153fff).rw(m_k052109, FUNC(k052109_device::read), FUNC(k052109_device::write)).umask(0x00ff).cswidth(16);
+	map(0x180000, 0x19ffff).ram().w(this, FUNC(gradius3_state::gfxram_w)).share("k052109");
 }
 
 
 void gradius3_state::gradius3_map2(address_map &map)
 {
 	map(0x000000, 0x0fffff).rom();
 	map(0x100000, 0x103fff).ram();
-	map(0x140000, 0x140001).w(this, FUNC(gradius3_state::cpuB_irqenable_w));
-	map(0x200000, 0x203fff).ram().share("share1");
-	map(0x24c000, 0x253fff).rw(this, FUNC(gradius3_state::k052109_halfword_r), FUNC(gradius3_state::k052109_halfword_w));
-	map(0x280000, 0x29ffff).ram().w(this, FUNC(gradius3_state::gradius3_gfxram_w)).share("k052109");
-	map(0x2c0000, 0x2c000f).rw(this, FUNC(gradius3_state::k051937_halfword_r), FUNC(gradius3_state::k051937_halfword_w));
-	map(0x2c0800, 0x2c0fff).rw(this, FUNC(gradius3_state::k051960_halfword_r), FUNC(gradius3_state::k051960_halfword_w));
-	map(0x400000, 0x5fffff).r(this, FUNC(gradius3_state::gradius3_gfxrom_r));     /* gfx ROMs are mapped here, and copied to RAM */
+	map(0x140000, 0x140000).w(this, FUNC(gradius3_state::cpuB_irqenable_w));
+	map(0x200000, 0x203fff).ram().share("shared_ram");
+	/* is this a bug in the game or something else? */
+	map(0x24c000, 0x253fff).rw(m_k052109, FUNC(k052109_device::read), FUNC(k052109_device::write)).umask(0x00ff).cswidth(16);
+	map(0x280000, 0x29ffff).ram().w(this, FUNC(gradius3_state::gfxram_w)).share("k052109");
+	map(0x2c0000, 0x2c000f).rw(this, FUNC(k051960_device::k051937_r), FUNC(k051960_device::k051937_w)).umask(0x00ff);
```

Comment:
can't use 'this' here and in the line below, you need to use the device tag
also 'umask' doesn't cut it, it's 'umask16' here and in the following lines